feat(language-service): add support for text replacement

ayazhafiz opened the pull request:
Adds a `replacementSpan` field on a completion that will allow typed
text to be replaced with the suggested completion value if a user
selects the completion. Previously, the completion value would simply be
appended to the text already typed. E.g. if we had

```
{{ti}}
```

typed in a template and `title` was recommended as a completion and
selected, the template would become

```
{{tititle}}
```

With `replacementSpan`, the original text `ti` will be replaced for
`title`.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

- [x] Feature

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

ayazhafiz commented:
This is WIP. I believe https://github.com/angular/angular/pull/31897 will need to land before this really works, because we are facing issues relating to invalid template AST sourceSpans (similar to what we saw, and still see, with the indexer module). In particular, I've included a test that fails with a `BoundEvent` template AST:

```typescript
@Component({
  template: `<a href="https://angular.io" (click)="han"></a>`,
})
export class AngularLinkComponent {
  handleClick() {}
}
```

Generating completions for `han` in the bound "click" event will produce an incorrect `replacementSpan` because the recorded span of the BoundEvent AST is incorrect.

ayazhafiz commented:
@andrius-pra this may interest you as well.

andrius-pra commented:
> Generating completions for `han` in the bound "click" event will produce an incorrect `replacementSpan` because the recorded span of the BoundEvent AST is incorrect.

Possible fix: You can add special case when node is ASTWithSource.
```typescript
let closestAst !: AstPath<AST>;
if (ast instanceof ASTWithSource) {
   closestAst = findAstAt(ast.ast, templatePosition - ast.sourceSpan.start);
} else {
   closestAst = findAstAt(ast, templatePosition - templateAstTail.sourceSpan.start.offset);
}
```

ayazhafiz commented:
> > Generating completions for `han` in the bound "click" event will produce an incorrect `replacementSpan` because the recorded span of the BoundEvent AST is incorrect.
> 
> Possible fix: You can add special case when node is ASTWithSource.
> 
> ```ts
> let closestAst !: AstPath<AST>;
> if (ast instanceof ASTWithSource) {
>    closestAst = findAstAt(ast.ast, templatePosition - ast.sourceSpan.start);
> } else {
>    closestAst = findAstAt(ast, templatePosition - templateAstTail.sourceSpan.start.offset);
> }
> ```

Good point. I'll try this out, but I think I did something similar in June and it wasn't very reliable. Thank you!

ayazhafiz committed:
feat(language-service): add support for text replacement

Adds a `replacementSpan` field on a completion that will allow typed
text to be replaced with the suggested completion value if a user
selects the completion. Previously, the completion value would simply be
appended to the text already typed. E.g. if we had

```
{{ti}}
```

typed in a template and `title` was recommended as a completion and
selected, the template would become

```
{{tititle}}
```

With `replacementSpan`, the original text `ti` will be replaced for
`title`.

ayazhafiz commented:
Okay, so half the problem seems to be solved -- we can now replace spans when performing a completion at the end of the text span! So in a case like `{{ti}}`, selecting a completion when typing after `i` will replace the entire span with the completion, like `{{title}}`.

Unfortunately, we cannot currently replace spans when selecting a completion in the middle of the  span. For example, say we had already typed `title` and then were typing it again, starting from the front:

```
{{tititle}}
    ^ typing position 
```

We'd expect that selecting the completion `title` would replace everything in the span, or at least `ti`. But it won't, because the way we currently grab the span to replace is by subtracting the size of the current span from the position that we are at.

One way to resolve this is to go work on the Angular AST even more, and have completely atomic ASTs for every semantic word or symbol in an Angular program. This would be a major overhaul and would take a lot of time, as we would have to do it for both expression ASTs and markup (HTML) ASTs. I think this is the most correct solution, but maybe not the most practical.

The other option is to perform some kind of matching algorithm - maybe we can figure out where the span to replace starts but looking at the overlap with the completion, then erasing the entire word in the span starting from the determined start location. I think for now this would work fine, but there is at least the issue that would not work with fuzzy-matched completions.

What do you think, @kyliau and @andrius-pra?

kyliau commented:
I think it's a "good-enough" solution with the absolute span. In edge cases like you mentioned, we could just set the replacement span to the nearest word boundary.

ayazhafiz commented:
Okay, maybe it's even safer to implement expanding the word until a word boundary rather than relying on a span. Let me see how that does.

ayazhafiz committed:
fixup! feat(language-service): add support for text replacement

ayazhafiz committed:
fixup! feat(language-service): add support for text replacement

ayazhafiz committed:
fixup! feat(language-service): add support for text replacement

ayazhafiz committed:
fixup! feat(language-service): add support for text replacement

andrius-pra commented on packages/language-service/src/completions.ts:
```diff
@@ -45,6 +45,69 @@ const ANGULAR_ELEMENTS: ReadonlyArray<ng.CompletionEntry> = [
+
+  // Expand on the left and right side until a word boundary is hit. Back up one expansion on both
+  // side to stay inside the word.
+  while (left >= 0 && templateSrc[left].match(WORD_PART)) --left;
```
It will not work for `|$any(...)` replacement. Consider to add $ character to WORD_PART or use `isIdentifierPart(code: number)` function from packages/compiler/src/expression_parser/lexer.ts.
```typescript
while (left >= 0 && isIdentifierPart(templateSrc.charCodeAt(left))) --left;
++left;
while (right < templateSrc.length && isIdentifierPart(templateSrc.charCodeAt(right))) ++right;
--right;
```

ayazhafiz replied:
It think it'd better to add `$` to `WORD_PART` rather than using `expression_parser/lexer` because we will be working on more than just expressions - markup can be replaced as well. Will fix this!

andrius-pra commented on packages/language-service/src/completions.ts:
```diff
@@ -110,7 +173,13 @@ export function getTemplateCompletions(
+  return result.map(entry => {
+    return {
+      ...entry,
+      replacementSpan: getBoundedWordSpan(templateInfo, position),
```
The `getBoundedWordSpan` method will always returns same result for all items. You can move outside `map` function.
```
  const replacementSpan = getBoundedWordSpan(templateInfo, position);
  return result.map(entry => ({...entry, replacementSpan}))
```

andrius-pra commented on packages/language-service/test/completions_spec.ts:
```diff
@@ -372,6 +372,156 @@ describe('completions', () => {
+    expect(completion).toBeDefined();
+    expect(completion.replacementSpan).toEqual({start: location.start - 3, length: 3});
+  });
+});
```
Can you also add tests for:
- [element name](https://github.com/andrius-pra/angular/commit/b2610650d090aa4ff1cc44219d361da6064d9cbe#diff-b43fa550ef00217368641108359ec9d4R523-R540): `<a~{elementName}pp-example></app-example>`
- [two-way binding](https://github.com/andrius-pra/angular/commit/b2610650d090aa4ff1cc44219d361da6064d9cbe#diff-b43fa550ef00217368641108359ec9d4R503-R521): `<div string-model [~{model}(model)]="field1"></div>`
- [names with _, $ characters](https://github.com/andrius-pra/angular/commit/b2610650d090aa4ff1cc44219d361da6064d9cbe#diff-b43fa550ef00217368641108359ec9d4R543-R562): `<div title="{{~{title}$field_1}}"></div>`

ayazhafiz committed:
fixup! feat(language-service): add support for text replacement

ayazhafiz committed:
fixup! feat(language-service): add support for text replacement

ayazhafiz committed:
fixup! feat(language-service): add support for text replacement

kyliau commented on packages/language-service/src/completions.ts:
```diff
@@ -45,6 +45,70 @@ const ANGULAR_ELEMENTS: ReadonlyArray<ng.CompletionEntry> = [
+  // A cursor is not itself a character in the template; it has a left (lower) and right (upper)
+  // index bound that hugs the cursor itself.
+  let templatePosition = position - template.span.start;
+  // To perform word expansion, we want to determine the left and right indeces that hug the cursor.
```
typo: indices

kyliau commented on packages/language-service/src/completions.ts:
```diff
@@ -45,6 +45,70 @@ const ANGULAR_ELEMENTS: ReadonlyArray<ng.CompletionEntry> = [
+ */
+function getBoundedWordSpan(templateInfo: AstResult, position: number): ts.TextSpan|undefined {
+  // Identifiers consist of alphanumeric characters, '_', or '$'.
+  const IDENTIFIER_PART = /[0-9a-zA-Z_$]/;
```
please make this a global outside of the function.

Since we are checking character by character, I wonder if it's more efficient to just check character code instead of a regex. Please see similar util functions here: https://github.com/angular/angular/blob/master/packages/compiler/src/chars.ts

(it's totally ok to leave this as-is, this is just a passing thought. We could worry about efficiency later.)

ayazhafiz replied:
Imported functions and performing character comparisons instead.

kyliau commented on packages/language-service/test/completions_spec.ts:
```diff
@@ -372,6 +372,184 @@ describe('completions', () => {
+  let ngLS: LanguageService;
+
+  beforeEach(() => {
+    mockHost = new MockTypescriptHost(['/app/main.ts']);
```
please use `mockHost.reset()` - it speeds up the test by a lot.

kyliau commented on packages/language-service/src/completions.ts:
```diff
@@ -45,6 +45,70 @@ const ANGULAR_ELEMENTS: ReadonlyArray<ng.CompletionEntry> = [
+  // 2. Case like
+  //      |rest of template
+  //    the cursor is at the start of the template, hugged only by the right side (0-index).
+  if (templatePosition === 0) {
```
Consider reorganizing the logic here a little for better readability

```
if (templatePosition === 0) {
  left = right = 0;
}
else if (templatePosition === templateSrc.length) {
  left = right = templateSrc.length - 1;
}
else {
  left = templatePosition - 1;
  right = templatePosition;
}
```

kyliau commented on packages/language-service/src/completions.ts:
```diff
@@ -45,6 +45,70 @@ const ANGULAR_ELEMENTS: ReadonlyArray<ng.CompletionEntry> = [
+  const IDENTIFIER_PART = /[0-9a-zA-Z_$]/;
+
+  const {template} = templateInfo;
+  const templateSrc = template.source;
```
Perhaps we should be a little paranoid here and bail if `templateSrc.length === 0`

kyliau commented on packages/language-service/test/completions_spec.ts:
```diff
@@ -372,6 +372,184 @@ describe('completions', () => {
+    const location = mockHost.getLocationMarkerFor(fileName, 'start');
+    const completions = ngLS.getCompletionsAt(fileName, location.start) !;
+    expect(completions).toBeDefined();
+    const completion = completions.entries.find(entry => entry.name === 'a') !;
```
where did `a` come from? The element `<a>`?
Maybe it's more helpful to also assert the completion kind.

ayazhafiz replied:
good point. I used a longer tag name and asserted the completion kind.

kyliau commented on packages/language-service/test/completions_spec.ts:
```diff
@@ -372,6 +372,184 @@ describe('completions', () => {
+    expect(completions).toBeDefined();
+    const completion = completions.entries.find(entry => entry.name === 'key') !;
+    expect(completion).toBeDefined();
+    expect(completion.replacementSpan).toEqual({start: location.start - 2, length: 5});
```
In the TS language service, the completion in this case would result in `obj.keykey`, so it seems like it's only looking to the left and never to the right. Do you think we should do the "right" thing (as was done here), or be consistent with TS autocomplete?

p/s: you can trigger autocomplete in vscode at any time by doing Ctrl+Space (on linux)
https://code.visualstudio.com/shortcuts/keyboard-shortcuts-linux.pdf

ayazhafiz replied:
Interesting. I would expect them to use the AST... but I guess not.

I think we are doing the correct and best thing. In general, I feel like it's preferable to exchange an entire word where that's the motivation. If we get  negative feedback from users, it will be easy to expand only to the left.

kyliau replied:
the "right" thing, pun intended. haha  😆

ayazhafiz replied:
haha, I like it!

kyliau commented:
Please also rebase this PR, thank you!

ayazhafiz committed:
fixup! feat(language-service): add support for text replacement

kyliau reviewed (approved):
LGTM, thank you!
